File: SimSky.m
function [soar_ss,Q,P,H,m2] = SimSky(A,W,m1,m2,k,c,query_vec,reorthog)

D=ApproDiag(A,c,k,W);

n=size(A,1);
final_vec_seq=zeros((k+1)*n,1);
[U,T,m1] = arnoldi(A,query_vec,m1,reorthog);
T_=T(1:m1,1:m1);
e_m_=sparse(1,1,1,m1,1);
beta=norm(T*T_^(k-1)*e_m_);
if beta < eps
    error('norm of the first element of vec_seq is %d', beta);
end
final_vec_seq(1:n)=D(:,1).* (U*T*T_^(k-1)*e_m_./beta);
final_vec_seq(n+1:2*n)=D(:,k+1).*(query_vec./beta);
for i=2:k
    final_vec_seq(i*n+1:(i+1)*n)=D(:,k+2-i).*(U*T*T_^(i-2)*e_m_./beta);
end

Q=zeros(n,m2+1);
P=zeros(n,m2+1);
H=zeros(m2+1,m2);
B=c*A';

V=zeros((k+1)*n,m2+1);
V(:,1)=final_vec_seq;
tol=n*eps;
for i=1:m2
    r=B*V(1:n,i)+V(k*n+1:(k+1)*n,i);
    t=V(1:k*n,i);
    s=[r;t];
    ow=norm(r);
    for j=1:i
        temp=r'*V(1:n,j);
        s=s-temp*V(:,j);
        H(j,i)=temp;
    end
    
    if reorthog==1
        for j=1:i
            temp=s(1:n)'*V(1:n,j);
            H(j,i)=H(j,i)+temp;
            s=s-temp*V(:,j);
        end
    end
    
    H(i+1,i)=norm(s(1:n));
    if H(i+1,i) < tol*ow
        m=i;
        H=H(1:1+m,1:m);
        V=V(:,1:1+m);
        Q=V(1:n,1:1+m);
        P=V(k*n+1:(k+1)*n,1:1+m);
    end
    V(:,i+1)=s/H(i+1,i);
end

Q=V(1:n,:);
P=V(k*n+1:(k+1)*n,:);

H1=H(1:m2,1:m2);
e_m=sparse(1,1,1,m2,1);
soar_ss=beta*Q*H*H1^(k-1)*e_m;

File: cppcode/SimSky.cpp
#include <Eigen/Sparse>
#include <unsupported/Eigen/SparseExtra>
#include <unsupported/Eigen/MatrixFunctions>
using namespace Eigen;

#include <iostream>
#include <chrono>
using namespace std;


#include "SimSky.h"
#include "Arnoldi.h"
#include "ApproDiag.h"

VectorXd SimSky(const SpM& mat, const MatrixXd& W, int m1, int m2, int k, double c, const SpV& query_vec, bool reorthog)
{
	const int n = mat.rows();
	MatrixXd V = MatrixXd::Zero((k + 1) * n, m2 + 1);
	MatrixXd H = MatrixXd::Zero(m2 + 1, m2);

	MatrixXd Q = MatrixXd::Zero(n, m2 + 1);
	MatrixXd P = MatrixXd::Zero(n, m2 + 1);

	// 	OrthoHess orth_hess = Arnoldi(mat, m1, query_vec, reorthog);


	auto start = chrono::system_clock::now();
	OrthoHess orth_hess = Arnoldi(mat, m1, query_vec, reorthog);
	auto end = chrono::system_clock::now();
	double dur = chrono::duration_cast<chrono::microseconds>(end - start).count();
	//cout << dur * chrono::microseconds::period::num / chrono::microseconds::period::den << endl;

	MatrixXd U_pre = orth_hess.U;
	MatrixXd T_pre = orth_hess.T;
	const int m1_pre = orth_hess.m;

	MatrixXd U_m1 = U_pre.block(0, 0, n, m1);
	MatrixXd T_m1 = T_pre.block(0, 0, m1, m1);



	// MatrixXd D = ApproDiag(mat, c, k, W);
	auto start1 = chrono::system_clock::now();
	MatrixXd D = ApproDiag(mat, c, k, W);
	auto end1 = chrono::system_clock::now();

	double dur1 = chrono::duration_cast<chrono::microseconds>(end1 - start1).count();

	//cout << dur1 * chrono::microseconds::period::num / chrono::microseconds::period::den << endl;

	SparseVector<double> e1(m1_pre);
	e1.setZero();
	e1.coeffRef(0) = 1;

	MatrixXd Tm_e1 = MatrixXd::Zero(m1, k);
	Tm_e1.col(0) = e1;

	for (int ii = 1; ii < k; ii++)
	{
		Tm_e1.col(ii) = T_m1 * Tm_e1.col(ii - 1);
	}

	MatrixXd U_Mul_T = U_pre * T_pre;
	VectorXd beta_vec = T_pre * Tm_e1.col(k - 1);
	double beta = beta_vec.norm();

	V.block(0, 0, n, 1) = 1 / beta * D.col(0).cwiseProduct(U_Mul_T * Tm_e1.col(k - 1));
	V.block(n, 0, n, 1) = 1 / beta * D.col(k).cwiseProduct(U_m1 * e1);

	MatrixXd U_T_Tm_e1 = MatrixXd::Zero(n, k - 1);

#pragma omp parallel for
	for (int jj = 0; jj < k - 1; jj++)
	{
		U_T_Tm_e1.col(jj) = U_Mul_T * Tm_e1.col(jj);

		V.block((jj + 2) * n, 0, n, 1) = 1 / beta * D.col(k - 1 - jj).cwiseProduct(U_T_Tm_e1.col(jj));
	}

	double eps = 1.0e-16;
	double tol = n * eps;

	for (int i = 0; i < m2; i++)
	{
		VectorXd V_i = V.col(i);
		VectorXd r = c * mat.transpose() * V_i.head(n) + V_i.tail(n);
		VectorXd t = V_i.head(k * n);
		VectorXd s((k + 1) * n);
		s.setZero();
		s << r, t;
		double ow = r.norm();
		//#pragma omp parallel for
		for (int j = 0; j <= i; j++)
		{
			VectorXd V_j = V.col(j);
			double temp = r.transpose() * V_j.head(n);
			s = s - temp * V_j;
			H(j, i) = temp;
		}
		if (reorthog == true) {
			for (int j = 0; j <= i; j++)
			{
				VectorXd V_j = V.col(j);
				double temp = s.head(n).transpose() * V_j.head(n);
				H(j, i) = H(j, i) + temp;
				s = s - temp * V_j;
			}
		}
		H(i + 1, i) = s.head(n).norm();
		if (H(i + 1, i) <= tol * ow)
		{
			m2 = i;
			H = H.block(0, 0, m2 + 1, m2);
			V = V.block(0, 0, (k + 1) * n, m2 + 1);
			Q = V.block(0, 0, n, m2 + 1);
			P = V.block(k * n, 0, n, m2 + 1);
		}
		V.col(i + 1) = s / H(i + 1, i);
	}
	Q = V.block(0, 0, n, m2 + 1);
	P = V.block(k * n, 0, n, m2 + 1);
	MatrixXd Hm = H.block(0, 0, m2, m2);

	SparseVector<double> e2(m2);
	e2.setZero();
	e2.coeffRef(0) = 1;

	MatrixPower<MatrixXd> Apow(Hm);
	MatrixXd Hm_power = Apow(k - 1);
	VectorXd Soar_ss = beta * Q * H * Hm_power * e2;
	return Soar_ss;
}